verifiers/internal/gcb/provenance.go

```diff
@@ -0,0 +1,281 @@
+	if err := self.isVerified(); err != nil {
+		return "", err
+	}
+
+	statement := self.verifiedIntotoStatementStruct
+	predicateBuilderID := statement.Predicate.Builder.ID
+
+	// Sanity check the builderID.
+	if err := isValidBuilderID(predicateBuilderID); err != nil {
+		return "", err
+	}
+
+	// Validate with user-provided value.
+	if builderOpts != nil && builderOpts.ExpectedID != nil {
+		if *builderOpts.ExpectedID != predicateBuilderID {
+			return "", fmt.Errorf("%w: expected '%s', got '%s'", serrors.ErrorMismatchBuilderID,
+				*builderOpts.ExpectedID, predicateBuilderID)
+		}
+	}
+
+	// Valiate that the recipe type is consistent.
+	if predicateBuilderID != statement.Predicate.Recipe.Type {
+		return "", fmt.Errorf("%w: expected '%s', got '%s'", serrors.ErrorMismatchBuilderID,
+			*builderOpts.ExpectedID, predicateBuilderID)
+	}
+
+	return predicateBuilderID, nil
+}
+
+func (self *GCBProvenance) VerifySubjectDigest(expectedHash string) error {
+	if err := self.isVerified(); err != nil {
+		return err
+	}
+
+	statement := self.verifiedIntotoStatementStruct
+	for _, subject := range statement.StatementHeader.Subject {
+		digestSet := subject.Digest
+		hash, exists := digestSet["sha256"]
+		if !exists {
+			return fmt.Errorf("%w: %s", serrors.ErrorInvalidDssePayload, "no sha256 subject digest")
+		}
+
+		if hash == expectedHash {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("expected hash '%s' not found: %w", expectedHash, serrors.ErrorMismatchHash)
+}
+
+// Verify source URI in provenance statement.
+func (self *GCBProvenance) VerifySourceURI(expectedSourceURI string) error {
+	if err := self.isVerified(); err != nil {
+		return err
+	}
+
+	statement := self.verifiedIntotoStatementStruct
+	materials := statement.Predicate.Materials
+	if len(materials) == 0 {
+		return fmt.Errorf("%w: no materials", serrors.ErrorInvalidDssePayload)
+	}
+	uri := materials[0].URI
+	if !strings.HasPrefix(expectedSourceURI, "https://") {
+		expectedSourceURI = "https://" + expectedSourceURI
+	}
+	if !strings.HasPrefix(uri, expectedSourceURI+"/commit/") {
+		return fmt.Errorf("%w: expected '%s', got '%s'",
+			serrors.ErrorMismatchSource, expectedSourceURI, uri)
+	}
+
+	return nil
+}
+
+func (self *GCBProvenance) VerifyBranch(branch string) error {
+	return fmt.Errorf("%w: GCB branch verification", serrors.ErrorNotSupported)
+}
+
+func (self *GCBProvenance) VerifyTag(tag string) error {
+	return fmt.Errorf("%w: GCB tag verification", serrors.ErrorNotSupported)
+}
+
+func (self *GCBProvenance) VerifyVersionedTag(tag string) error {
+	return fmt.Errorf("%w: GCB versioned-tag verification", serrors.ErrorNotSupported)
+}
+
+func (self *GCBProvenance) VerifySignature() error {
+	if len(self.gcloudProv.ProvenanceSummary.Provenance) == 0 {
+		return fmt.Errorf("%w: no provenance found", serrors.ErrorInvalidDssePayload)
+	}
+	// Assume a single provenance in the array.
+	prov := self.gcloudProv.ProvenanceSummary.Provenance[0]
```

Comment:
TODO: need to iterate. This will support multiple version specs in the future (v0.1, v0.2, v1.0, etc).